### PR TITLE
Plumb firewall rules

### LINKED ISSUE #1017: Send current firewall rules applying to a guest in the sled agent instance-creation request
#955 adds initial integration with OPTE, which provides the virtual networks over which guests communicate. That pull request implements the core integration, but isn't fully complete end-to-end. In particular, guests in the same VPC Subnet still can't directly communicate with one another, since the default firewall disposition in OPTE is to block all such traffic. Developers can get around this by adding firewall rules to OPTE with a command line invocation like:

```
bnaecker@feldspar : ~/omicron $ pfexec /opt/oxide/opte/bin/opteadm add-fw-rule -p opte0 --action allow --dir in --hosts subnet=172.30.0.0/22 --ports any --priority 10 --protocol tcp
```

where `opte0` is the guest OPTE port object, and the subnet matches whatever the guest was created with.

This issue tracks providing the current set of firewall rules that apply to a guest in the actual instance-creation request sent to the sled agent. This is distinct from the work required for distributing a _new_ set of firewall rules to the right sleds when the rules are modified via the public API, it's only the initial rules that apply at creation time.

### PR DESCRIPTION
**Pull request by plotnick**
Plumb VPC firewall rules through Nexus → sled-agent → OPTE. Builds on @bnaecker's work in the [propagate-firewall-rules-to-sled branch](https://github.com/oxidecomputer/omicron/compare/main...propagate-firewall-rules-to-sled), which in turn was based on work started in PR #465.

Outstanding issues:
- [x] ~Target interfaces are matched by MAC address only right now. Is this ok?~ Use (VNI, MAC) pair to match.
- [x] IP and network targets are not yet supported. ~The console doesn't support these yet either, but the API does.~
- [x] Send default rules to sled agent on instance creation. Right now rules are only sent on VPC creation and rule updates. Fixes #1017.
- [ ] Missing integration tests (issue #1792). I have tested this manually with multiple instances and a small variety of targets and host filters, but we definitely need automated tests.
- [x] Fix existing tests, which are now failing with `cross-VPC firewall target unsupported`

### CODE REVIEW

**Review by bnaecker — COMMENTED** [review]
This is a lot of great work, thanks for taking it on! I see your TODO-list on the PR comment itself. Is the plan to tackle those now as part of this PR? Or are those going to be separate issues?

File: nexus/src/app/vpc.rs
Comment (plotnick) [reply nexus/src/app/vpc.rs]: 1dc02d3b now raises an error for these cases, which should never occur (if they do, it's because we made a mistake in specifying `DEFAULT_FIREWALL_RULES`).

File: nexus/src/app/vpc.rs
Comment (plotnick) [reply nexus/src/app/vpc.rs]: Fixed in 1dc02d3b.

File: nexus/src/app/vpc.rs
Comment (plotnick) [reply nexus/src/app/vpc.rs:540]: Excellent question; 1fa97ab8 does explicit de-duplication on the NICs. (If `sled_agent_client::types::NetworkInterface` were `Hash`, this could be simplified by accumulating directly into a `HashSet`; please LMK if you know how to do that.)

File: nexus/src/db/datastore/network_interface.rs
Comment (plotnick) [reply nexus/src/db/datastore/network_interface.rs:279]: Hairy indeed, but probably worth it. Refactoring of these three routines forthcoming.

File: nexus/src/db/datastore/vpc.rs
Comment (plotnick) [reply nexus/src/db/datastore/vpc.rs]: Done in 578a2d9f, thanks.

File: sled-agent/src/opte/illumos/port_manager.rs
Comment (plotnick) [reply sled-agent/src/opte/illumos/port_manager.rs]: Great, that's what I thought.

File: sled-agent/src/opte/illumos/port_manager.rs
Comment (plotnick) [reply sled-agent/src/opte/illumos/port_manager.rs]: Fixed in 1994a43d, thanks.

File: nexus/src/db/datastore/network_interface.rs
Comment (plotnick) [reply nexus/src/db/datastore/network_interface.rs]: Ok, factored using a boxed query in 902e6a84.

**Comment by rzezeski** [comment]
Heads up: connectivity inbound to a guest will fail when firewall rules are populated until oxidecomputer/opte#252 lands.

**Comment by rzezeski** [comment]
Okay, oxidecomputer/opte#252 is fixed. So inbound connectivity should work with latest xde. If it doesn't please let me know.

File: nexus/src/db/datastore/vpc.rs
Comment (bnaecker) [inline nexus/src/db/datastore/vpc.rs:710]: This function name and / or comment could be a bit clearer. Maybe `resolve_vpc_subnets_to_ip_networks`? Since the word "subnet" is overloaded, I've generally found it useful to be really pedantic when they're both in play.

File: nexus/src/db/datastore/vpc.rs
Comment (bnaecker) [inline nexus/src/db/datastore/vpc.rs:714]: Out of curiosity, why a hash map rather than a B-tree? `Name` is sortable, and they should all be unique since they're scoped to the VPC.

File: sled-agent/src/opte/illumos/port_manager.rs
Comment (bnaecker) [inline sled-agent/src/opte/illumos/port_manager.rs:317]: Is the extra scope here necessary?

File: sled-agent/src/opte/illumos/port_manager.rs
Comment (bnaecker) [inline sled-agent/src/opte/illumos/port_manager.rs:558]: These seem to be a sequence of pretty straightforward type conversions, e.g., converting a list of ports in the control plane representation into that for OPTE. I think it might help readability (and allow unit testing) if we factor these into `From` implementations. Then this becomes a list of expressions like `let action = VpcFirewallRuleAction::from(rule.action);` etc.

Comment (bnaecker) [reply sled-agent/src/opte/illumos/port_manager.rs:558]: This really applies to the whole closure, actually.

File: sled-agent/src/opte/illumos/port_manager.rs
Comment (bnaecker) [inline sled-agent/src/opte/illumos/port_manager.rs:511]: Note to our future selves -- we may just want to not send disabled rules.

File: nexus/src/app/vpc.rs
Comment (bnaecker) [inline nexus/src/app/vpc.rs:10]: We've been trying to move to item-per-line imports to aid conflict resolution.

Comment (bnaecker) [reply nexus/src/app/vpc.rs:10]: There are a couple of other places below with grouped imports as well, lines 32 and 36.

File: nexus/src/app/vpc.rs
Comment (bnaecker) [inline nexus/src/app/vpc.rs:73]: Doesn't look like this clone is needed, but I could be missing a reference to `vpc` below.

File: nexus/src/app/vpc.rs
Comment (bnaecker) [inline nexus/src/app/vpc.rs:351]: I think the comment could be a bit more explanatory. At first this appeared to just return the default rules, but I (think) it's actually inserting the real VPC name in place of the name `"default"`. If that's correct, that'd be very helpful to have in the comment.

File: nexus/src/app/vpc.rs
Comment (bnaecker) [inline nexus/src/app/vpc.rs:430]: Is there an issue to track this? I think it's a pretty big and multifaceted work-stream, since it's a potentially long-running, retry-style operation.

File: nexus/src/app/vpc.rs
Comment (bnaecker) [reply nexus/src/app/vpc.rs:518]: I think you can implement `Hash` manually, probably by hashing all the fields. Not sure if it's worth it, up to you at this point, since I don't want to further delay this PR unnecessarily.

**Review by bnaecker — COMMENTED** [review]
My apologies for the long delay. There's a lot of great work here, thank you! I've mostly got questions about improving names or comments, but the functionality itself looks great.

File: nexus/src/app/sagas/instance_migrate.rs
Comment (plotnick) [reply nexus/src/app/sagas/instance_migrate.rs:232]: No problem; linked in `a8a6d3ea`. I didn't see any issues for any of the other related `TODO`s, so I made issue #1783 track all of them; LMK if you think we need individual issues immediately.

And thanks for linking to #1017; I think this PR closes that, right?

Comment (bnaecker) [reply nexus/src/app/sagas/instance_migrate.rs:232]: Yep, it does indeed!

File: nexus/src/db/datastore/vpc.rs
Comment (plotnick) [reply nexus/src/db/datastore/vpc.rs]: Good suggestion, done in 8ab92afb.

File: nexus/src/app/vpc.rs
Comment (plotnick) [reply nexus/src/app/vpc.rs]: Sure thing, fixed in `f3b64ae0`.

File: nexus/src/db/datastore/vpc.rs
Comment (plotnick) [reply nexus/src/db/datastore/vpc.rs]: Nope, no good reason. Changed to use `BTreeMap` in cbedb56b.

File: sled-agent/src/opte/illumos/port_manager.rs
Comment (plotnick) [reply sled-agent/src/opte/illumos/port_manager.rs]: Not really, no; it was just that the `rules` aren't used anywhere else, but probably a premature optimization. Removed in adc62ccc.

File: sled-agent/src/opte/illumos/port_manager.rs
Comment (plotnick) [reply sled-agent/src/opte/illumos/port_manager.rs]: So.... I like this idea very much, and had originally intended to use it; these conversions are by far my least favorite part of this PR. The problem is Rust's "orphan rule": `only traits defined in the current crate can be implemented for types defined outside of the crate`. So the question becomes: where should such `From` implementations live? We pretty clearly don't want `oxide-vpc` to depend on `omicron-common`, and it's not obvious to me that we'd want the converse dependency, either. (But maybe it's fine?)

Another option would be to use strings as an intermediate representation, as most of the `oxide-vpc` types have a `FromStr` implementation, and most of the `omicron-common` types have a `Display` implementation. But I think in general it's a bad idea to do that, and would much rather use manual conversions.

Commit 9869832a attempts to take a middle ground: it refactors the `opte_firewall_rules` routine so that it uses a new (local) `FromVpcFirewallRule` trait that includes methods for all of the component conversions. It also moves it into a new `firewall_rules` module so that it's not cluttering up `port_manager`. LMK if you think this is better/worse/good enough. I think it's cleaner than it was, in particular since the basic logic of `opte_firewall_rules` is less obscured by the component conversions, but I'm open to further refactoring if you think it's warranted.

File: nexus/src/app/vpc.rs
Comment (plotnick) [reply nexus/src/app/vpc.rs]: Good catch, I wonder why clippy didn't complain? Anyway, fixed in f7ba1eb7.

File: nexus/src/app/vpc.rs
Comment (plotnick) [reply nexus/src/app/vpc.rs]: Correct, added in 171a6251.

**Comment by plotnick** [comment]
@bnaecker: Thanks for the very helpful second review! The one remaining concern I have that I'd like to fix is that if you use a non-existent VPC name in a target or host filter, things currently fail in a rather opaque way. I'm going to try to fix that today (Wed), and then I think it'll be in ok shape (though see also issue #1756, which I think is serious but not blocking).

However, I'd like to maybe delay actually merging this until @smklein branches for the demo next week. Since this change involves touching many critical pieces (including instance creation), does not have integration tests, and is not necessary for that demo, I think it's a bit risky to land with so little time for testing. Other opinions welcome.

File: sled-agent/src/opte/illumos/port_manager.rs
Comment (bnaecker) [reply sled-agent/src/opte/illumos/port_manager.rs]: I think what you have there is a solid solution, yeah. If `From` isn't possible due to orphan rules (ugh, I forgot about that), I usually go for a standalone conversion function. Since these are all POD types with no private data, that's pretty straightforward. The trait solution you have is great as well, nice idea!

**Comment by bnaecker** [comment]
That's all excellent, thanks for chasing down those threads @plotnick! I think your plan of merging after the demo branch is created sounds very wise. I'm approving this now, feel free to merge whenever that happens. And again, nice work. It's really excellent to see such a core networking primitive start to take shape!

**Review by bnaecker — APPROVED** [review]

**Pull request merged**
